beacon-chain/sync/pending_attestations_queue.go

```diff
@@ -72,162 +79,295 @@ func (s *Service) processPendingAttsForBlock(ctx context.Context, bRoot [32]byte
 }
 
 func (s *Service) processAttestations(ctx context.Context, attestations []any) {
-	for _, signedAtt := range attestations {
-		// The pending attestations can arrive as both aggregates and attestations,
-		// and each form has to be processed differently.
-		switch t := signedAtt.(type) {
+	if len(attestations) == 0 {
+		return
+	}
+
+	atts := make([]ethpb.Att, 0, len(attestations))
+	for _, att := range attestations {
+		switch v := att.(type) {
 		case ethpb.Att:
-			s.processAtt(ctx, t)
+			atts = append(atts, v)
 		case ethpb.SignedAggregateAttAndProof:
-			s.processAggregate(ctx, t)
+			s.processAggregate(ctx, v)
 		default:
-			log.Warnf("Unexpected item of type %T in pending attestation queue. Item will not be processed", t)
+			log.Warnf("Unexpected attestation type %T, skipping", v)
 		}
 	}
+
+	for _, bucket := range s.bucketAttestationsByData(atts) {
+		s.processAttestationBucket(ctx, bucket)
+	}
 }
 
-func (s *Service) processAggregate(ctx context.Context, aggregate ethpb.SignedAggregateAttAndProof) {
-	att := aggregate.AggregateAttestationAndProof().AggregateVal()
+// attestationBucket groups attestations with the same AttestationData for batch processing.
+type attestationBucket struct {
+	dataHash     [32]byte
+	data         *ethpb.AttestationData
+	attestations []ethpb.Att
+}
 
-	// Save the pending aggregated attestation to the pool if it passes the aggregated
-	// validation steps.
-	valRes, err := s.validateAggregatedAtt(ctx, aggregate)
-	if err != nil {
-		log.WithError(err).Debug("Pending aggregated attestation failed validation")
-	}
-	aggValid := pubsub.ValidationAccept == valRes
-	if s.validateBlockInAttestation(ctx, aggregate) && aggValid {
-		if features.Get().EnableExperimentalAttestationPool {
-			if err = s.cfg.attestationCache.Add(att); err != nil {
-				log.WithError(err).Debug("Could not save aggregated attestation")
-				return
-			}
+// bucketAttestationsByData groups attestations by their AttestationData hash.
+func (s *Service) bucketAttestationsByData(attestations []ethpb.Att) map[[32]byte]*attestationBucket {
```

Comment:
This doesn't need to be pointer receiver? 
```suggestion
func bucketAttestationsByData(attestations []ethpb.Att) map[[32]byte]*attestationBucket {
```